packages/completer-extension/src/index.ts

```diff
@@ -45,337 +41,175 @@ namespace CommandIDs {
- */
-const consoles: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:consoles',
-  requires: [ICompletionManager, IConsoleTracker],
+const defaultProvider: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlab/completer-extension:base-service',
+  requires: [ICompletionProviderManager],
   autoStart: true,
   activate: (
     app: JupyterFrontEnd,
-    manager: ICompletionManager,
-    consoles: IConsoleTracker
+    serviceManager: ICompletionProviderManager
   ): void => {
-    // Create a handler for each console that is created.
-    consoles.widgetAdded.connect((sender, widget) => {
-      const anchor = widget.console;
-      const editor = anchor.promptCell?.editor ?? null;
-      const session = anchor.sessionContext.session;
-      // TODO: CompletionConnector assumes editor and session are not null
-      const connector = new CompletionConnector({ session, editor });
-      const handler = manager.register({ connector, editor, parent: widget });
-
-      const updateConnector = () => {
-        const editor = anchor.promptCell?.editor ?? null;
-        const session = anchor.sessionContext.session;
-
-        handler.editor = editor;
-        // TODO: CompletionConnector assumes editor and session are not null
-        handler.connector = new CompletionConnector({ session, editor });
-      };
-
-      // Update the handler whenever the prompt or session changes
-      anchor.promptCellCreated.connect(updateConnector);
-      anchor.sessionContext.sessionChanged.connect(updateConnector);
-    });
-
-    // Add console completer invoke command.
-    app.commands.addCommand(CommandIDs.invokeConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.invoke, { id });
-        }
-      }
-    });
-
-    // Add console completer select command.
-    app.commands.addCommand(CommandIDs.selectConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.select, { id });
-        }
-      }
-    });
-
-    // Set enter key for console completer select command.
-    app.commands.addKeyBinding({
-      command: CommandIDs.selectConsole,
-      keys: ['Enter'],
-      selector: `.jp-ConsolePanel .jp-mod-completer-active`
-    });
+    serviceManager.registerProvider(new ContextCompleterProvider());
+    serviceManager.registerProvider(new KernelCompleterProvider());
   }
 };
 
-/**
- * An extension that registers notebooks for code completion.
- */
-const notebooks: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:notebooks',
-  requires: [ICompletionManager, INotebookTracker],
+const manager: JupyterFrontEndPlugin<ICompletionProviderManager> = {
+  id: COMPLETION_MANAGER_PLUGIN,
+  requires: [
+    INotebookTracker,
```

Comment:
Do we need all the trackers to be specified as `requires`?

Before they were split into separate plugins so downstream lab-based distributions could cherry-pick the plugins they wanted.

So maybe we can make them `optional` here now.

Reply:
Indeed, these trackers are removed in my second PR https://github.com/jupyterlab/jupyterlab/commit/625e8ec03aef73b35f22cd75e42c0f31b93f949f